CLW/CL/CLW.cl

```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
         }
     }
 }
+
+// --------------------- ATOMIC OPERTIONS ------------------------
+
+#define DEFINE_ATOMIC(operation)\
+    inline void atomic_##operation##_float(volatile __global float* addr, float value)\
+    {\
+        union{\
+        unsigned int u32;\
+        float        f32;\
+        } next, expected, current;\
+        current.f32 = *addr;\
+        do{\
+            expected.f32 = current.f32;\
+            next.f32 = operation(expected.f32, value);\
+            current.u32 = atomic_cmpxchg((volatile __global unsigned int *)addr,\
+                expected.u32, next.u32);\
+        } while (current.u32 != expected.u32);\
+    }
+
+#define DEFINE_ATOMIC_FLOAT3(operation)\
+    inline void atomic_##operation##_float3(volatile __global float3* addr, float3 value)\
+    {\
+        volatile __global float* p = (volatile __global float*)addr;\
+        atomic_##operation##_float(p, value.x);\
+        atomic_##operation##_float(p + 1, value.y);\
+        atomic_##operation##_float(p + 2, value.z);\
+    }
+
+inline void atomic_max_int(volatile __global int* addr, int value)
+{
+    atomic_max(addr, value);
+}
+
+inline void atomic_min_int(volatile __global int* addr, int value)
+{
+    atomic_min(addr, value);
+}
+
+// --------------------- HELPERS ------------------------
+
+#define DEFINE_ASSIGN_OPERATOR(type)\
+    inline void assign_##type(__local type* addr, type value)\
+    {\
+        *addr = value;\
+    }
+
+inline void assign_float3(__local float3* addr, float3 value)
+{
+    (*addr).xyz = value.xyz;
+}
+
+inline int divide_int(int dividend, int divider)
+{
+    return dividend / (divider != 0 ? divider : 1);
+}
+
+inline float divide_float(float dividend, float divider)
+{
+    return dividend / (fabs(divider) > epsilon ? divider : 1.f);
+}
+
+inline float3 divide_float3(float3 dividend, float3 divider)
+{
+    return (float3)(divide_float(dividend.x, divider.x),
+                    divide_float(dividend.y, divider.y),
+                    divide_float(dividend.z, divider.z));
+} 
+
+// --------------------- REDUCTION ------------------------
+
+#define DEFINE_REDUCTION(bin_op, type)\
+__kernel void reduction_##bin_op##_##type(__global type* buffer,\
+                                          int buf_count,\
+                                          __local type* shared_mem,\
+                                          __global type* out)\
+{\
+    int global_id = get_global_id(0);\
+    int group_id = get_group_id(0);\
+    int local_id = get_local_id(0);\
+    int group_size = get_local_size(0);\
+    \
+    if (global_id < buf_count)\
+    {\
+        assign_##type(shared_mem + local_id, buffer[global_id]);\
+    }\
+    else\
+    {\
+        assign_##type(shared_mem + local_id, neutral_##bin_op##_##type);\
+    }\
+    \
+    if (global_id == 0)\
+    {\
+        *out = neutral_##bin_op##_##type;\
```

Comment:
What if global_id == 0 is executed last and you are doing atomic additions from other threads on non-initialized memory?

Reply:
Rewrited